### ISSUE TITLE
Allow other Attribute names instead of [GenerateShape]

### ISSUE THREAD

**Issue by Tornhoof** [eiriktsarpalis/typeshape-csharp#1]
As the preferred way to identify types for srcgen is via Attributes, the amount of Marker Attributes on each type is kinda getting out of hand. I have some types with 4-5 marker attributes just for srcgen on them (cloning, BFS/DFS, mapping etc). So I wonder is there a way to use only one Marker Attribute for everything.

I'm aware of the approach to use a marker type, which is then annotated with the types which should be identified, but maybe there is another solution.

I'm not even certain it works, but maybe msbuild ``CompilerVisibleProperty`` can help there, i.e. put the fully qualified Attribute name in a specific msbuild property and use that Name as part of the logic.

**Comment by eiriktsarpalis** [comment]
> I have some types with 4-5 marker attributes just for srcgen on them (cloning, BFS/DFS, mapping etc). So I wonder is there a way to use only one Marker Attribute for everything.

What this library is trying to show fundamentally is that you needn't have to attach bespoke attributes for every application, instead one general-purpose attribute marking a type as "traversable" should be enough. Some source generators address this problem using interceptors, but the TypeShape generator is general-purpose and cannot recognize application-specific callsites.

> i.e. put the fully qualified Attribute name in a specific msbuild property and use that Name as part of the logic.

This sounds a bit fragile to me when taking refactoring into consideration.

**Comment by Tornhoof** [comment]
> This sounds a bit fragile to me when taking refactoring into consideration.

Now I'm curious, why do you think this is fragile or rather more fragile than the whole srcgen approach.
Yes you'd have some magic strings, but you have a lot of those already in srcgens, they are usually full of const string for type names.
At least from my experience, if refactoring goes wrong, you'll rapidly see compilation errors and that's a plus compared to either reflection or expression trees.

Yes I'm aware that your library reduces the fragility of the srcgen approach, as a lot of common code is done by it. As far as I can see, I can probably replace 2-3 srcgens with your approach.

**Comment by eiriktsarpalis** [comment]
> Yes you'd have some magic strings, but you have a lot of those already in srcgens, they are usually full of const string for type names.

But if you rename-refactor your types, that should result in a well-behaved source generator producing valid up-to-date source code. If you're referring to magic strings inside source generator implementations, it's true that some of those are unavoidable but those should be constrained to APIs that are part of the runtime components invoked by the source generator -- if not, that sounds like a bug in the generator.

**Comment by Tornhoof** [comment]
> But if you rename-refactor your types, that should result in a well-behaved source generator producing valid up-to-date source code. 

Yes, but how would putting an attribute Name in a specific msbuild property in the callsite's csproj be fragile then? The only thing I can think of is moving the callsite's attribute to a different namespace or renaming it, then you need to update the property in your callsite's csproj or build props.

**Comment by eiriktsarpalis** [comment]
Oh, I mistook your requirement as specifying the fully qualified names of the generated types, but you're saying that the marker attribute itself should be customizable.

**Comment by Tornhoof** [comment]
Yeah, just the marker attribute, I apologize if was not clear enough in my description.
The goal is to cut down on the amount of marker attributes, by reusing marker attribute names of other srcgens.

**Comment by eiriktsarpalis** [comment]
It should be doable in principle, as an additive property perhaps. Would you be interested in contributing a PoC?

**Comment by Tornhoof** [comment]
Yes, I'll see what I can get working tomorrow

**Comment by Tornhoof** [comment]
I have a working POC at PR #2 

This is way more involved than anticipated, due to the fact that Incremental Source Generators parse options from msbuild into ``IncrementalValueProviders`` too. The trick here is to use a bag-style type ``MarkerAttributeHolder`` to hold the parsed msbuild property data during pipeline execution. So it's not easy to get the attribute name early in the pipeline.
I was not aware of that when I started, simply because I looked at the cookbook for the non-incremental source generators.


The whole bag-style type stinks of breaking the incremental pipeline imho, but I couldn't verify that.

I copied the full ``ForTypesWithAttributeDeclaration`` method to ``ForTypesWithOptionalAttributeDeclaration``  which then uses that ``MarkerAttributeHolder`` object for the attribute name etc.

Both the original ``IncrementalValuesProvider`` and the optional marker attribute one are than concatenated prior to the parsing logic in ``ParseFromGenerateShapeAttributes``

I apologize for the formatting noise in some of the files in the pr.

Bottom line: experiment works, patient is dead. Imho this too complicated and some "unclean" tricks in code paths which are performance sensitive.

**Comment by eiriktsarpalis** [comment]
> Bottom line: experiment works, patient is dead. Imho this too complicated and some "unclean" tricks in code paths which are performance sensitive.

Thanks. I actually started reviewing before I caught your comments here. We can revisit if you feel confident that it can be improved.

> unrelated: Those ultra long tuple/valuetuple tests apparently generate too long file paths (on my system atleast)

I generally enable long paths on my windows machines, so hadn't encountered that issue. I'll see if I can add a note in the README.
